File: java/social-api/src/main/java/org/apache/shindig/social/dataservice/PersonHandler.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package org.apache.shindig.social.dataservice;

import org.apache.shindig.social.ResponseError;
import org.apache.shindig.social.ResponseItem;
import org.apache.shindig.social.opensocial.model.Person;

import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import com.google.inject.Inject;

import java.util.Set;


public class PersonHandler extends DataRequestHandler {
  private PersonService personService;

  @Inject
  public PersonHandler(PersonService personService) {
    this.personService = personService;
  }

  protected ResponseItem handleDelete(RequestItem request) {
    return new ResponseItem<Object>(ResponseError.BAD_REQUEST,
        "You can't delete people. ", null);
  }

  protected ResponseItem handlePut(RequestItem request) {
    return new ResponseItem<Object>(ResponseError.NOT_IMPLEMENTED,
        "You can't add people right now. ", null);  }

  protected ResponseItem handlePost(RequestItem request) {
    return new ResponseItem<Object>(ResponseError.NOT_IMPLEMENTED,
        "You can't add people right now. ", null);
  }

  /**
   * /people/{userId}/{groupId}/{optionalPersonId}
   *
   * examples:
   * /people/john.doe/@all
   * /people/john.doe/@friends
   * /people/john.doe/@self
   */
  protected ResponseItem handleGet(RequestItem request) {
    String[] segments = getParamsFromRequest(request);

    UserId userId = UserId.fromJson(segments[0]);
    GroupId groupId = GroupId.fromJson(segments[1]);

    String optionalPersonId = null;
    if (segments.length > 2) {
      optionalPersonId = segments[2];
    }

    if (optionalPersonId != null
        || groupId.getType() == GroupId.Type.self) {
      return personService.getPerson(userId, request.getToken());
    }

    PersonService.SortOrder sort = getEnumParam(request, "orderBy",
        PersonService.SortOrder.topFriends, PersonService.SortOrder.class);
    PersonService.FilterType filter = getEnumParam(request, "filterBy",
        PersonService.FilterType.all, PersonService.FilterType.class);

    int first = getIntegerParam(request, "startIndex", 0);
    int max = getIntegerParam(request, "count", 20);

    Set<String>  profileDetails = Sets.newHashSet(
        getListParam(request, "fields",
            Lists.newArrayList(Person.Field.ID.toString(),
                Person.Field.NAME.toString(),
                Person.Field.THUMBNAIL_URL.toString())));

    return personService.getPeople(userId, groupId, sort, filter, first, max,
        profileDetails, request.getToken());
  }

}

File: php/src/socialrest/PersonHandler.php
<?php
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

class PersonHandler extends DataRequestHandler {
	private $service;

	public function __construct()
	{
		$service = Config::get('people_service');
		$this->service = new $service();
	}

	/**
	 * /people/{userId}/{groupId}/{optionalPersonId}
	 *
	 * examples:
	 * /people/john.doe/@all
	 * /people/john.doe/@friends
	 * /people/john.doe/@self
	 */
	public function handleGet($params, $token)
	{
		$userId = UserId::fromJson($params[1]);
		$groupId = GroupId::fromJson($params[2]);
		$optionalPersonId = false;
		if (count($params) > 3) {
			$optionalPersonId = $params[3];
		}
		$fields = !empty($_GET['fields']) ? explode(',', $_GET['fields']) : null;
		if ($optionalPersonId || $groupId->getType() == 'self') {
			//FIXME same logic as the java code here, but doesn't seem to do much with the optionalPersonId which seems odd
			return $this->service->getPerson($userId, $groupId, $fields, $token);
		}
		$sort = !empty($_GET['orderBy']) && in_array($_GET['orderBy'], PeopleService::$sortOrder) ? $_GET['orderBy'] : '';
		$filter = !empty($_GET['filterBy']) && in_array($_GET['filterBy'], PeopleService::$filterType) ? $_GET['filterBy'] : '';
		$first = !empty($_GET['startIndex']) && is_numeric($_GET['startIndex']) ? $_GET['startIndex'] : 0;
		$max = !empty($_GET['count']) && is_numeric($_GET['count']) ? $_GET['count'] : 0;
		return $this->service->getPeople($userId, $groupId, $sort, $filter, $first, $max, $fields, $token);
	}

	public function handleDelete($params, $token)
	{
	    return new ResponseItem(BAD_REQUEST, "You can't delete people.", null);
	}
	
	public function handlePost($params, $token)
	{
	    return new ResponseItem(NOT_IMPLEMENTED, "You can't edit people right now.", null);
	}

	public function handlePut($params, $token)
	{
	    return new ResponseItem(NOT_IMPLEMENTED, "You can't add people right now.", null); 
	}
}